### DISCUSSION TITLE
[Bug] pip install tts fails

### DISCUSSION THREAD

**Discussion by pombears** [coqui-ai/TTS#2100] [General]
### pip install tts failing on Windows
On Windows, running "pip install tts" fails with below errors in output:

> Building wheels for collected packages: tts
>   Building wheel for tts (pyproject.toml) ... error
>   error: subprocess-exited-with-error
> 
>   × Building wheel for tts (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [1203 lines of output]

....

>  building 'TTS.tts.utils.monotonic_align.core' extension
>       error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for tts
> Failed to build tts
> ERROR: Could not build wheels for tts, which is required to install pyproject.toml-based projects

Full pip install output:
[full-pip-output.txt](https://github.com/coqui-ai/TTS/files/9765258/full-pip-output.txt)

I already have the following installed on my machine, so looks like the above error may be misleading?
Microsoft Visual C++ 2015-2022 Redistributable (x64) - 14.31.31103
Microsoft Visual C++ 2015-2022 Redistributable (x86) - 14.30.30704

Attempts made to resolve issue:

- Downloaded and installed the Microsoft C++ Build Tools
- Using Visual Studio Installer (gets installed from previous step), installed Visual Studio Community 2022
- In Visual Studio Installer, added Workload "Desktop Development with C++"

I'm now getting error:
 

> error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for tts
>   **Building wheel for pyworld (setup.py) ... error**
>   **error: subprocess-exited-with-error**
> 
>   × **python setup.py bdist_wheel did not run successfully.**
>   │ exit code: 1
>   ╰─> [12 lines of output]
>       C:\Users\anthony\AppData\Roaming\Python\Python310\site-packages\setuptools\dist.py:771: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
>         warnings.warn(
>       running bdist_wheel
>       running build
>       running build_py
>       creating build
>       creating build\lib.win-amd64-cpython-310
>       creating build\lib.win-amd64-cpython-310\pyworld
>       copying pyworld\__init__.py -> build\lib.win-amd64-cpython-310\pyworld
>       running build_ext
>       building 'pyworld.pyworld' extension
>       error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for pyworld
>   Running setup.py clean for pyworld
> **Failed to build tts pyworld**
> **ERROR: Could not build wheels for tts, which is required to install pyproject.toml-based projects**
> 

--------------------------------

### To Reproduce

Open command prompt
Type and enter: pip install tts

### Expected behavior

TTS should install without error

### Environment

```shell
Windows 10 (Version 21H2 - OS Build 19044.2130)
Python 3.10.6
PIP version 22.2.2
```

**Comment by thorstenMueller** [comment]
I had some struggles with pyworld dependency on Windows too. But i found a solution and show how to get it up and running here:
https://youtu.be/zRaDe08cUIk?t=743

Maybe this helps you.

**Comment by pombears** [comment]
Hi Thorsten

Thanks very much for this. From googling around I wonder if it's a case of needing to run it in  a virtual environment running Python 3.8 rather than my default 3.10. I will give that link a try and see how I go.

**Comment by Vuizur** [comment]
I had problems installing it on Windows as well, what I did at the end was using poetry to create a virtual environment with python 3.9, because python 3.10 doesn't seem to work, and then installing the package from git (where pyworld is not a necessary dependency anymore).

**Comment by pombears** [comment]
Hi Vuizur. Thanks for this. I'm curious to know if you needed to do anything else beyond installing the git repository inside a 3.8 virtual environment to get it to work.

**Comment by Vuizur** [comment]
No, it worked fine after that (I used 3.9 though). 👍 If you should be stuck at any step, just ask.

**Comment by pombears** [comment]
Thanks Vuizur. I've just created a virtual environment running Python v3.9 but no joy currently.
1. Created a container folder "C:\Projects"
2. Downloaded the git repository into this folder using below (successful): 
**git clone https://github.com/coqui-ai/TTS.git**
3. In VS Code, I opened the TTS subfolder created by the download
4. In a terminal window in VS Code, ran the following to configure the TTS folder as my virtual environment running Python v3.9 specifically:
**& "C:\Program Files\Python39\python.exe" -m venv .**                           (successful)
5. With current working directory still at the TTS folder, I then ran the following to install the contents of the git repository:
**\Scripts\pip install -e .[all,dev,notebooks]**

Currently getting below errors. I've also attached full pip install output.
[pip_install_output.txt](https://github.com/coqui-ai/TTS/files/9820169/pip_install_output.txt)

> Building wheels for collected packages: TTS
>   Building editable for TTS (pyproject.toml) ... error
>   error: subprocess-exited-with-error
> 
>   × Building editable for TTS (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [87 lines of output]
>       <string>:34: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.

and further down.....

 

>  File "C:\Users\anthony\AppData\Local\Temp\pip-build-env-tnknhu1y\overlay\Lib\site-packages\setuptools\_distutils\_msvccompiler.py", line 423, in compile
>           raise CompileError(msg)
>       distutils.errors.CompileError: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.33.31629\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
>       error: Support for editable installs via PEP 660 was recently introduced
>       in `setuptools`. If you are seeing this error, please report to:
>      
>       https://github.com/pypa/setuptools/issues
>      
>       Meanwhile you can try the legacy behavior by setting an
>       environment variable and trying to install again:
>      
>       SETUPTOOLS_ENABLE_FEATURES="legacy-editable"
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building editable for TTS
> Failed to build TTS
> ERROR: Could not build wheels for TTS, which is required to install pyproject.toml-based projects
> WARNING: You are using pip version 22.0.4; however, version 22.3 is available.
> You should consider upgrading via the 'C:\Projects\TTS\Scripts\python.exe -m pip install --upgrade pip' command.

**Comment by Vuizur** [comment]
What I recommend is installing [Poetry](https://python-poetry.org/docs/), then executing:
```powershell
poetry new tts-test
cd tts-test
```
Then download Python 3.9 (or 3.8) and set poetry's python version to the older version like [explained here.](https://stackoverflow.com/questions/60580113/change-python-version-to-3-x)
Then run
```
poetry add git+https://github.com/coqui-ai/TTS
```
Then run `poetry shell` and synthesize speech like [here](https://tts.readthedocs.io/en/latest/inference.html), so for example by running `tts-server` to get the interactive gui (you only need to open the link in the console in the browser.)

(I wrote this out of memory, so I hope it works).

**Comment by pombears** [comment]
Thanks Vuizur. I will give it a try. Prior to running "poetry new tts-test", how should I actually install Poetry on Windows.
There is a script at https://install.python-poetry.org/ which I can download, save as say poetry-installer.py and then run potentially run that using my existing Python v3.9 executable:
"C:\Program Files\Python39\python.exe" poetry-installer.py

Would that be the way to go, or is there another way I should be installing it?

P.S. I don't suppose it matters, but the specific version of Python v3.9 I installed in the existing virtual environment is v3.9.13 (the highest version of v.39 python.org is offering as an installable at the moment)

**Comment by Vuizur** [comment]
I think the best way is to run the Windoes powershell install command in powershell `(Invoke-WebRequest -Uri https://install.python-poetry.org -UseBasicParsing).Content | py -`

**Comment by pombears** [comment]
Okay I've installed poetry using the PowerShell command method, added poetry bin directory to PATH variable, tested running "**poetry --version**" does return the version number of poetry.
I then created my virtual environment in C:\Projects by running "**poetry new tts-test**" in this location.
That part was okay.
But when then going into the tts-test folder and then running "poetry env use 3.9" I get the following error:

```
c:\Projects\tts-test>poetry env use 3.9
'python3.9' is not recognized as an internal or external command,
operable program or batch file.

Command python3.9 -c "import sys; print(sys.executable)" errored with the following return code 1, and output:

(blank)
```

**Comment by Vuizur** [comment]
I think you need the absolute path, which you get with `py -0p`

**Comment by pombears** [comment]
Okay. This is what I get:

```
c:\Projects\tts-test>py -0p
Installed Pythons found by py Launcher for Windows
 -3.10-64       C:\Program Files\Python310\python.exe *
 -3.9-64        C:\Program Files\Python39\python.exe


c:\Projects\tts-test>poetry env use C:\Program Files\Python39\python.exe

Too many arguments to "env use" command, expected arguments "python"
```

**Comment by Vuizur** [comment]
Can you try it with "C:\Program Files\Python39\python.exe" with quote marks?

**Comment by pombears** [comment]
Good point! My bad!
Getting the below now. My main installed python version is v3.10 but need to make my tts-test use 3.9
```
c:\Projects\tts-test>poetry env use "C:\Program Files\Python39\python.exe"

The specified Python version (3.9.13) is not supported by the project (^3.10).
Please choose a compatible version or loosen the python constraint specified in the pyproject.toml file.
```

**Comment by Vuizur** [comment]
You need to edit the pyproject.toml to (change 3.10 to 3.9)

**Comment by pombears** [comment]
Thank you. Getting further now...

```
c:\Projects\tts-test>poetry env use "C:\Program Files\Python39\python.exe"
Creating virtualenv tts-test-mKFQEpZO-py3.9 in C:\Users\anthony\AppData\Local\pypoetry\Cache\virtualenvs
Using virtualenv: C:\Users\anthony\AppData\Local\pypoetry\Cache\virtualenvs\tts-test-mKFQEpZO-py3.9

c:\Projects\tts-test>poetry add git+https://github.com/coqui-ai/TTS

Updating dependencies
Resolving dependencies... (0.0s)

The current project's Python requirement (>=3.9,<4.0) is not compatible with some of the required packages Python requirement:
  - tts requires Python >=3.7.0, <3.11, so it will not be satisfied for Python >=3.11,<4.0

Because tts-test depends on tts (0.8.0) @ git+https://github.com/coqui-ai/TTS@HEAD which requires Python >=3.7.0, <3.11, version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties

    For tts, a possible solution would be to set the `python` property to ">=3.9,<3.11"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers

```

**Comment by pombears** [comment]
Using ">=3.9,<3.11" now and it's download packages

**Comment by pombears** [comment]
The packages seemed to install okay, but at the end I'm getting the same kind of error I got when using venv and VS Code:
```

    "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -IC:\Users\anthony\AppData\Local\Temp\pip-build-env-hh4savuo\overlay\Lib\site-packages\numpy\core\include -IC:\Users\anthony\AppData\Local\pypoetry\Cache\virtualenvs\tts-test-mKFQEpZO-py3.9\include "-IC:\Program Files\Python39\include" "-IC:\Program Files\Python39\Include" "-IC:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\VS\include" /TcTTS/tts/utils/monotonic_align/core.c /Fobuild\temp.win-amd64-cpython-39\Release\TTS/tts/utils/monotonic_align/core.obj    core.c
    C:\Program Files\Python39\include\pyconfig.h(59): fatal error C1083: Cannot open include file: 'io.h': No such file or directory
    error: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.33.31629\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
    [end of output]

    note: This error originates from a subprocess, and is likely not a problem with pip.
    ERROR: Failed building wheel for TTS
  Failed to build TTS
  ERROR: Could not build wheels for TTS, which is required to install pyproject.toml-based projects


  at ~\AppData\Roaming\pypoetry\venv\lib\site-packages\poetry\utils\env.py:1476 in _run
      1472│                 output = subprocess.check_output(
      1473│                     command, stderr=subprocess.STDOUT, env=env, **kwargs
      1474│                 )
      1475│         except CalledProcessError as e:
    → 1476│             raise EnvCommandError(e, input=input_)
      1477│
      1478│         return decode(output)
      1479│
      1480│     def execute(self, bin: str, *args: str, **kwargs: Any) -> int:

The following error occurred when trying to handle this error:


  PoetryException

  Failed to install C:/Users/anthony/AppData/Local/pypoetry/Cache/virtualenvs/tts-test-mKFQEpZO-py3.9/src/TTS

  at ~\AppData\Roaming\pypoetry\venv\lib\site-packages\poetry\utils\pip.py:51 in pip_install
       47│
       48│     try:
       49│         return environment.run_pip(*args)
       50│     except EnvCommandError as e:
    →  51│         raise PoetryException(f"Failed to install {path.as_posix()}") from e
       52│

```

**Comment by pombears** [comment]
Further up the following sticks out:

 ```
• Installing unidic-lite (1.0.8)
  • Installing tts (0.8.0 ef82adc): Failed   <<<<<<<<<<<<<<<<<

  CalledProcessError
```

**Comment by Vuizur** [comment]
Have you tried the solution under https://stackoverflow.com/questions/40018405/cannot-open-include-file-io-h-no-such-file-or-directory ?

**Comment by pombears** [comment]
No not yet. I will definitely give that a go. Thanks very much for the help Vuizur. Almost here

**Reply by Grandmother001** [reply]
Did you have any luck getting this to work?

**Comment by erogol** [comment]
Thanks for the issue but w e don't support win and currently we don't intend to. So I move this to the discussions.

**Comment by vivisol** [comment]
I have fixed the problem that  when trying `pip install TTS` ocurred error  `Failed to build tts
ERROR: Could not build wheels for tts, which is required to install pyproject.toml-based projects`.

# Step 1: 
download TTS package from `pypi.org`, unzip the file `TTS-0.11.1.tar.gz`

# Step 2:
run `pip install -r requirements.txt`

# Step 3:
run `python setup.py install `，
then it probably will occur error ` building 'TTS.tts.utils.monotonic_align.core' extension
error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/`.
Don't worry, just do it as it suggests, download Microsoft C++ build tools on website https://visualstudio.microsoft.com/visual-cpp-build-tools/.
![image](https://user-images.githubusercontent.com/22468237/220297144-8835a4f3-e9fe-4638-8027-eeea02b05c30.png)

# Step 4:
After installing the Microsoft C++ build tools ,  run `python setup.py install`, then it succeed.

# Step 5:
run `pip list` ,then you can see TTS has been successfully installed.
![image](https://user-images.githubusercontent.com/22468237/220298070-815c161e-56eb-4d52-b01b-0c126027a524.png)

**Reply by b-rania** [reply]
Thank you for the tutorial! It seemed to work for me !

**Reply by adiitea** [reply]
I'm still facing same problem even after following all the steps

**Reply by eginhard** [reply]
@adiitea Well if you don't share the command you ran and the error you got you won't get any help... (and the steps here are very outdated)

**Reply by adiitea** [reply]
I want to install the storyteller-core package...and that need other version of numpy and tts needs other version so idk what to do that's why I dropped the idea of working on that project

**Reply by eginhard** [reply]
@adiitea So I already wrote here two times that you should use [our fork](https://github.com/idiap/coqui-ai-TTS) because this repo is abandoned and broken. Please follow the installation instructions there and if you still have problems open an issue there with more details. I'll stop replying here now.

**Reply by adiitea** [reply]
I used that also but still tts is not showing 
Pip show tts is giving package not found.

**Reply by lexkoro** [reply]
Did you run "pip install coqui-tts"?

**Reply by adiitea** [reply]
Yesss

**Reply by adiitea** [reply]
Seee i want to install storyteller-core package and for that I need to install tts but it's not installing ....I'm developing a ai-story generator

**Comment by Ifromfridaynightfunk** [comment]
came from Microsoft Support team (thanks for them), And is there a way to use this, as i'm confused and don't know any shit about coding and all of that, especially on Windows 11, I also have an issue where i can't install and TTS Packages, If possible can you help me?

**Comment by tomsepe** [comment]
try installing Visual Studio Build Tools 2022 [Microsoft C++ Build Tools ](https://visualstudio.microsoft.com/visual-cpp-build-tools/) and make sure you check the box for the "Desktop development with C++" Install option

**Comment by kaisch13** [comment]
doesn't work for python 3.12 since "trainer" is not updated on this python version yet:.(

**Comment by dvl-harish** [comment]
If anyone still facing the issue, you need to install Microsoft C++ build tools with Python 3.10. I have tested and it is working fine.

**Comment by eginhard** [comment]
[Our fork](https://github.com/idiap/coqui-ai-TTS) now provides prebuilt wheels for Windows, so the C++ Build Tools shouldn't be needed anymore. You can install it with `pip install coqui-tts` (Python 3.12 is supported as well there)

**Reply by adiitea** [reply]
Still getting the error

**Comment by adiitea** [comment]
I'm still facing this error pip install -r requirements.txt is now also giving error relate to cython "getting requirements to build wheel does not run successfully"

**Reply by eginhard** [reply]
As written above, you should use [our fork](https://github.com/idiap/coqui-ai-TTS) (available via `pip install coqui-tts`), which also works fine on Windows. This repo is abandoned and doesn't work with recent Python and dependency versions anymore.

**Comment by hazardland** [comment]
Solved installation after attempting to install it for 2 days : D

1. First you will need to find Microsoft Build Tools in case you are wondering what is minimal installation of its components for installing TTS here you go:

![image](https://github.com/user-attachments/assets/12b45eae-cd6e-4cd5-909e-69f01088e386)

2. After which this two commands work with:

Python 3.10.11
Cython 0.29.37

```
pip install Cython==0.29.37
pip install TTS --no-build-isolation
```

But consider for the moment writing this comment the secret is in Cython==0.29.37, take a close look it is 37, not lower version, lower versions are failing, also higher version after 0.29.37 is is 3.x.x which is also failing

**Comment by Calciumcarbonate321** [comment]
This was fixed in a recent PR in pip (https://github.com/pypa/pip/pull/13001).

Install it by doing `pip install git+https://github.com/pypa/pip.git`
